feat(package): Enable replica set for the MongoDB results cache and configure it when starting the package.

junhaoliao opened the pull request:
<!--
Set the PR title to a meaningful commit message that:
- follows the Conventional Commits specification (https://www.conventionalcommits.org).
- is in imperative form.
Example:
fix: Don't add implicit wildcards ('*') at the beginning and the end of a query (fixes #390).
-->

# Description
<!-- Describe what this request will change/fix and provide any details necessary for reviewers. -->
1. Enable replica set for the MongoDB results cache.
2. Configure replica set in the indices building script.
3. Change the results cache's Docker network mode to `host` and specify `--bind-ip` and `--port` when starting the mongo image.

# Validation performed
<!-- Describe what tests and validation you performed on the change. -->
## Search query performance compassion
With & Without the replica set enabled, performed a search query in the WebUI with a screen recording of 30fps, and extract below times:
1. T0 (relative 0 second): the query was started.
4. T1: the first batch of results were available and displayed in the browser.
5. T2: the total number of results are available (either from the results-metadata or summed from the aggregated time buckets) and displayed in the browser.
6. T3: the job is marked as done and displayed in the browser.

  | T0 | T1 | T2 | T3
-- | -- | -- | -- | --
Trial 1 (with replica-set) | 0 | 0.7 | 1.266667 | 2.1
Trial 2 (with replica-set) | 0 | 0.633333 | 1.433333 | 2.166667
Trial 3 (with replica-set) | 0 | 0.433333 | 1.4 | 2.1
Trial 4 (w/o replica-set) | 0 | 0.466667 | 1.433333 | 2.1
Trial 5 (w/o replica-set) | 0 | 0.5 | 1.433333 | 2.066667
Trial 6 (w/o replica-set) | 0 | 0.6 | 1.433333 | 2.2

![image](https://github.com/user-attachments/assets/54043ddb-2956-4721-ba6b-02e8ccb56189)

The speed performance does not seem to significantly differ with and without replica-set enabled.

## Result cache host switching
Previous attempts to enable replica set failed because there were host reachability issues observed. Below validations ensure the `clp-config.yml`/results_cache host and port are configurable without affecting the result cache's avaialbility.

 `build/clp-package/lib/python3/site-packages/clp_py_utils/create-results-cache-indices.py`

0. Built the clp-package.
1. `clp-package/sbin/start-clp.sh` with the default config and observed below logs:
    ```
    2024-12-25T22:45:58.892 INFO [start_clp] Creating results_cache indices...
    2024-12-26 03:45:59,548 [DEBUG] Replica set initialization requested for localhost:27017
    2024-12-26 03:45:59,551 [DEBUG] Replica set has not been previously initialized.
    2024-12-26 03:45:59,551 [DEBUG] Initializing replica set at localhost:27017
    2024-12-26 03:45:59,693 [DEBUG] Replica set initialized successfully.
    2024-12-25T22:46:00.517 INFO [start_clp] Created results_cache indices.
    ```
2. Changed the `result_cache`'s `host` and `port` config in `clp-package/etc/clp-config.yml` to be `localhost:27018`. e.g.,
    ```yaml
    results_cache:
      host: "localhost"
      port: 27018
      db_name: "clp-query-results"
      stream_collection_name: "stream-files"
    ```

    `clp-package/sbin/start-clp.sh` and observed below logs:
    ```
    2024-12-25T22:50:29.309 INFO [start_clp] Creating results_cache indices...
    2024-12-26 03:50:29,761 [DEBUG] Replica set initialization requested for localhost:27018
    2024-12-26 03:50:30,266 [DEBUG] Replica set is already initialized but reports invalid config.
    2024-12-26 03:50:30,266 [DEBUG] Initializing replica set at localhost:27018
    2024-12-26 03:50:30,269 [DEBUG] Replica set initialized successfully.
    2024-12-25T22:50:31.042 INFO [start_clp] Created results_cache indices.
    ```
4. Used `ip addr` to find another address that belongs to the clp-package host (i got `172.22.78.194`). Changed the `result_cache`'s `host` and `port` config in `clp-package/etc/clp-config.yml` to be `172.22.78.194:27018`. `clp-package/sbin/start-clp.sh` and observed below logs:
    ```
    2024-12-25T22:54:43.400 INFO [start_clp] Creating results_cache indices...
    2024-12-26 03:54:43,919 [DEBUG] Replica set initialization requested for 172.22.78.194:27018
    2024-12-26 03:54:44,422 [DEBUG] Replica set is already initialized but reports invalid config.
    2024-12-26 03:54:44,422 [DEBUG] Initializing replica set at 172.22.78.194:27018
    2024-12-26 03:54:44,427 [DEBUG] Replica set initialized successfully.
    2024-12-25T22:54:45.239 INFO [start_clp] Created results_cache indices.
    ```
7. Changed the `result_cache`'s `host` and `port` config in `clp-package/etc/clp-config.yml` to be `localhost:27017`. `clp-package/sbin/start-clp.sh` and observed below logs:
    ```
    2024-12-25T22:57:01.151 INFO [start_clp] Creating results_cache indices...
    2024-12-26 03:57:01,619 [DEBUG] Replica set initialization requested for localhost:27017
    2024-12-26 03:57:02,122 [DEBUG] Replica set is already initialized but reports invalid config.
    2024-12-26 03:57:02,122 [DEBUG] Initializing replica set at localhost:27017
    2024-12-26 03:57:02,125 [DEBUG] Replica set initialized successfully.
    2024-12-25T22:57:02.894 INFO [start_clp] Created results_cache indices.
    ```

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208964046828607

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Configuration Changes**
	- Updated MongoDB configuration to support replica sets
	- Removed explicit IP binding settings
	- Added replica set name configuration

- **MongoDB Management**
	- Introduced new functions for managing MongoDB replica set initialization and status checking

- **Deployment Updates**
	- Modified container startup scripts for results cache and reducer
	- Updated network configuration settings for containers
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

junhaoliao committed:
Enable MongoDB replica set for the results cache and initialize it when starting the package.

junhaoliao commented:
An issue is found when in `clp-config.yml` I configure a non-27017 port on host `localhost` for the replica set:
```
pymongo.errors.OperationFailure: No host described in new configuration with {version: 36921, term: -1} for replica set rs0 maps to this node, full error: {'ok': 0.0, 'errmsg': 'No host described in new configuration with {version: 36921, term: -1} for replica set rs0 maps to this node', 'code': 74, 'codeName': 'NodeNotFound'}
```
However, such issue is only observed on `localhost` hosts. If the host is not `localhost` / `127.0.0.1` (e.g., my WSL's virtual eth0 interface gets assigned `172.22.78.194`), with any port the replica set initialization would succeed and there is no issue connecting to the clp-configured mongo address.

By looking into the community Mongo server source code, it was found they have two ways to check whether a node is up before the server configures itself as a replica node: [`isSelfFastPath`](https://github.com/mongodb/mongo/blob/c79f785a7cc5c304d49bab06e0bb1e722d1ec968/src/mongo/db/repl/isself.cpp#L196) and [`isSelfSlowPath`](https://github.com/mongodb/mongo/blob/c79f785a7cc5c304d49bab06e0bb1e722d1ec968/src/mongo/db/repl/isself.cpp#L258).
* `isSelfFastPath` is purely logic based as it tries to check identity of the server's address with a given config's host and port. It first checks if the port matches then it checks for host.
* `isSelfSlowPath` handles the case that the fast path cannot find a match of the server's address in any of the configs. It tries to make a connection with the given config host and port and issue a command to check for identity.

### Why `localhost`: `any non-27017 port` fails
Since our mongo server instance runs within a Docker container with only the docker's localhost's port 27017 mapped to the Docker host, if we request replica set to be configured on `localhost:27017`, the fast path check should succeed; if we request for `localhost`: `any non-27017 port`, the fast path wouldn't find a match because the server is started with 27017, and the slow path wouldn't find a match because in the Docker container's own network, the Docker host's port (configured by clp-config and used in the replica set config) is not reachable.

### Why `any non-localhost host`: `any port` works
The fast path would fail but the slow path would succeed as Docker knows `non-localhost` is outside of its network and Docker routes the connection to the Docker host.

## Proposed solution
1. Create a docker network and put all services (docker containers) that use Mongo into the same network.
2. `docker run` the result cache with `--network "host"` so that the mongo Docker container sees / exposes everything on the Docker host.

Considering existing CLP utilities, Option 2. seems the quickest way. @haiqi96 may comment on the impacts on CLP cloud.

junhaoliao commented:
You may also see from the `{version: 36921, term: -1}` in the error message that they may have some uninitialized memory issue, lol

junhaoliao committed:
Enable host network mode and specify `--bind_ip` and `--port` when starting `mongo`.

junhaoliao committed:
Add new replica set configuration logic.

junhaoliao committed:
Add back empty line.

haiqi96 commented:
Discussed with Junhao offline and clarified the two points:
1. The current replica set only adds one member, so it doesn't actually provide redundancy.
2. The purpose of replica set is not support redundancy, but to enable oplog that supports watch(), which replaces polling to check updates in the collection.

Imo, we should make these two points clear in the code. We have agreed on renaming some replica related functions and adding docstrings so developer who only reads code can also have the context above.

haiqi96 commented on components/clp-package-utils/clp_package_utils/scripts/start_clp.py:
```diff
@@ -491,6 +491,7 @@ def start_results_cache(instance_id: str, clp_config: CLPConfig, conf_dir: pathl
     cmd = [
         "docker", "run",
         "-d",
+        "--network", "host",
```
Just curious, do you know exactly why we need --network host? is there any IP other than results_cache.port that will be accessed?

junhaoliao commented on components/clp-py-utils/clp_py_utils/create-results-cache-indices.py:
shall we also consider renaming this script as `initialize-results-cache.py` to better reflect its updated purpose?

junhaoliao committed:
Merge branch 'main' into results-cache-replica-set

junhaoliao committed:
Add docs; Rename functions & parameters to clarify their purpose and improve readability.

junhaoliao committed:
Rename results cache initialization script.

junhaoliao committed:
Docs nit.

junhaoliao commented:
@haiqi96 I added docstrings for the newly added functions in the init script, and renamed the methods & parameters & the init script name, which hopefully has clarified the purpose of those methods. Please help taking a look again.

haiqi96 commented on components/clp-py-utils/clp_py_utils/initialize-results-cache.py:
```diff
@@ -0,0 +1,146 @@
+
+    :return: A tuple containing:
+        - is_already_initialized: Indicates whether the replica set has already been initialized.
+        - should_configure_replica_set: Indicates whether the replica set needs to be configured or
+          reconfigured.
```
The sentence reads to me as if the should_configure_replica_set = bool means needs to be configured and should_configure_replica_set = false means needs to be reconfigured.
How about we do this:

```suggestion
        - should_configure_replica_set: Indicates whether the replica set needs to be configured (or
          reconfigured).
```

haiqi96 commented on components/clp-py-utils/clp_py_utils/initialize-results-cache.py:
```diff
@@ -0,0 +1,146 @@
+def check_replica_set_status(client: MongoClient, netloc: str) -> tuple[bool, bool]:
+    """
+    Checks the current replica set status of the MongoDB server and determines whether it needs to
+    be configured or reconfigured.
```
ditto.

junhaoliao replied:
```suggestion
    be configured (or reconfigured).
```

haiqi96 commented on components/clp-py-utils/clp_py_utils/initialize-results-cache.py:
```diff
@@ -0,0 +1,146 @@
+    be configured or reconfigured.
+
+    :param client:
+    :param netloc: The network location for which the replica set configuration is being checked.
```
How do you think about?

The network location of the replica set configuration

It's simpler but not sure if it's accurate enough.

junhaoliao replied:
sure
```suggestion
    :param netloc: The network location of the replica set configuration.
```

haiqi96 commented on components/clp-py-utils/clp_py_utils/initialize-results-cache.py:
```diff
@@ -0,0 +1,146 @@
+    :param client:
+    :param netloc: The network location of the MongoDB instance to configure
+    as a replica set member.
+    :param force: Whether this is a reconfiguration operation.
```
How about:
```suggestion
    :param force: Forces reconfiguration.
```

haiqi96 reviewed (commented):
Left a few comments on the docstrings but in general the change looks reasonable to me.

junhaoliao committed:
Docs - Apply suggestions from code review

Co-authored-by: haiqi96 <14502009+haiqi96@users.noreply.github.com>

junhaoliao committed:
Put "or reconfigured" into parentheses in docs.

junhaoliao committed:
Docs - Apply suggestions from code review

haiqi96 reviewed (approved):
Looks good to me, assuming it is tested and verified.
